src/main/scala/org/graphframes/GraphFrame.scala

```diff
@@ -265,6 +266,85 @@ class GraphFrame private(
     edges.select(explode(array(SRC, DST)).as(ID)).groupBy(ID).agg(count("*").cast("int").as("degree"))
   }
 
+  // ========================= Partition By ====================================
+  val PARTITION_ID: String = "partition_id"
+
+  /**
+   * A [[org.apache.spark.Partitioner]] that use the key of PairRDD as partition
+   * id number.
+   */
+  class ExactAsKeyPartitioner(partitions: Int) extends Partitioner {
+    require(partitions >= 0, s"Number of partitions ($partitions) cannot be negative.")
+
+    def numPartitions: Int = partitions
+
+    def getPartition(key: Any): Int = {
+      val partitionIdAsKey = key.asInstanceOf[Int]
+      return partitionIdAsKey
+    }
+  }
+
+  /**
+   * Implements a watered down version of Graphx partitionBy.
+   *
+   * @param numPartitions Number of partitions to be split by
+   * @param strategy any case object of Graphx's PartitionStrategy trait
+   *
+   * @return A new `GraphFrame` constructed from new edges and existing vertices
+   */
+  def partitionBy(numPartitions: Int, strategy: PartitionStrategy): GraphFrame = {
+    val getPartitionIdUdf = udf[PartitionID, Long, Long, Int] {
+      (src, dst, numParts) => strategy.getPartition(src, dst, numParts)
+    }
+
+    // Remove 'src' and 'dst' and get original 'attr' cols
+    val (unnestedAttrCols, _) = edgeColumnMap.filter { p => 
+        val key = p._1
+        key != SRC && key != DST
+      }
+      .toSeq.sortBy(_._2)
+      .unzip
+
+    //  Construct the flatten columns of edges + new partition id col
+    val edgesWithPartitionIdColumns = Seq(
+      Seq(col(SRC), col(DST)),
+      unnestedAttrCols.map(c => col(ATTR + "." + c)),
+      Seq(col(PARTITION_ID))).flatten
```

Comment:
is it possible to create a `Seq` without having to flatten one?

Reply:
I will use scala.collection.mutable.ListBuffer[Column] instead, then toSeq. My bad on this one.